### PR TITLE
CNV_1020 Live and Readiness Probes

### PR DESCRIPTION
**Pull request by bgaydosrh**
See Jira story: https://jira.coreos.com/browse/CNV-1020

Created new assembly titled **cnv-monitoring-vm-health**. 
Created module titled **cnv-configure-liveness-readiness-probes**.
Updated Topic Map to add **Monitoring virtual machine health** topic.

Test Build: http://file.bos.redhat.com/bgaydos/111219/cnv/cnv_users_guide/cnv-monitoring-vm-health.html 

Peer review requested from @sheriff-rh @ahardin-rh or @adellape 

Code review requested from Mark Sluiter via Jira and email.

Label **Peer Review Needed** and **enterprise-4.3**.

Many thanks!

Bob

### CODE REVIEW

File: modules/cnv-configure-liveness-readiness-probes.adoc
Comment (ahardin-rh) [inline modules/cnv-configure-liveness-readiness-probes.adoc]: s/gives you/provides

File: modules/cnv-configure-liveness-readiness-probes.adoc
Comment (ahardin-rh) [inline modules/cnv-configure-liveness-readiness-probes.adoc]: s/configures/configure

File: modules/cnv-configure-liveness-readiness-probes.adoc
Comment (ahardin-rh) [inline modules/cnv-configure-liveness-readiness-probes.adoc]: s/configures/configure

**Comment by ahardin-rh** [comment]
@bgaydosrh Just some minor comments from me. Otherwise, looks good!

**Comment by bgaydosrh** [comment]
> @bgaydosrh Just some minor comments from me. Otherwise, looks good!

Thanks Ashley, I appreciate your promptness and help :-)

Bob

File: modules/cnv-configure-liveness-readiness-probes.adoc
Comment (slintes) [inline modules/cnv-configure-liveness-readiness-probes.adoc]: this is a very specific URL from our dev environment. Better just use `image: kubevirt/fedora-cloud-registry-disk-demo`.

File: modules/cnv-configure-liveness-readiness-probes.adoc
Comment (slintes) [inline modules/cnv-configure-liveness-readiness-probes.adoc]: same

File: modules/cnv-configure-liveness-readiness-probes.adoc
Comment (slintes) [inline modules/cnv-configure-liveness-readiness-probes.adoc]: same

File: modules/cnv-configure-liveness-readiness-probes.adoc
Comment (slintes) [inline modules/cnv-configure-liveness-readiness-probes.adoc]: on the rendered page I see `:leveloffset!:` at the end, whatever that means, but I think it doesn't belong there :)

**Review by slintes — COMMENTED** [review]
only some small nits, otherwise lgtm

File: modules/cnv-configure-liveness-readiness-probes.adoc
Comment (bgaydosrh) [reply modules/cnv-configure-liveness-readiness-probes.adoc]: Done.

File: modules/cnv-configure-liveness-readiness-probes.adoc
Comment (bgaydosrh) [reply modules/cnv-configure-liveness-readiness-probes.adoc]: Done.

File: modules/cnv-configure-liveness-readiness-probes.adoc
Comment (bgaydosrh) [reply modules/cnv-configure-liveness-readiness-probes.adoc]: Done.

File: modules/cnv-configure-liveness-readiness-probes.adoc
Comment (bgaydosrh) [reply modules/cnv-configure-liveness-readiness-probes.adoc]: Thanks. Replaced 3 instances of this, one in each section.

File: modules/cnv-configure-liveness-readiness-probes.adoc
Comment (bgaydosrh) [reply modules/cnv-configure-liveness-readiness-probes.adoc]: See above. Done.

File: modules/cnv-configure-liveness-readiness-probes.adoc
Comment (bgaydosrh) [reply modules/cnv-configure-liveness-readiness-probes.adoc]: See above. Done.

File: modules/cnv-configure-liveness-readiness-probes.adoc
Comment (bgaydosrh) [reply modules/cnv-configure-liveness-readiness-probes.adoc]: Ah, thanks! My end code block (----) was not aligned properly. Done. Will update link for test build shortly.

**Comment by bgaydosrh** [comment]
Code and peer review complete.

@alexxa can you QE review this for me? 

Once that occurs I can merge and pick to 4.2 and 4.3, if no objections from @slintes 

Thanks everyone,
Bob

**Comment by alexxa** [comment]
@bgaydosrh I can't validate it right now because of no 4.3 environment and a blocker to create one. It should be resolved in a day or two, I hope.

**Comment by bgaydosrh** [comment]
> @bgaydosrh I can't validate it right now because of no 4.3 environment and a blocker to create one. It should be resolved in a day or two, I hope.

Gotcha. No problem, @alexxa.

File: cnv/cnv_users_guide/cnv-monitoring-vm-health.adoc
Comment (stu-gott) [inline cnv/cnv_users_guide/cnv-monitoring-vm-health.adoc]: This is missing a comma? "... which back Services**,** "

I might be inclined to re-phrase this sentence.

If readiness probes fail, the `VirtualMachineInstance` will be removed from applicable Endpoints until the probe recovers.

**Review by stu-gott — COMMENTED** [review]
technical content looks good.

File: cnv/cnv_users_guide/cnv-monitoring-vm-health.adoc
Comment (bgaydosrh) [reply cnv/cnv_users_guide/cnv-monitoring-vm-health.adoc]: Thanks @stu-gott --- yes, that's much better. Something struck me as not quite right about this either, but without the clarity of the intention it escaped me. The way I had it worded,  Endpoints was the subject but as the which clause is not needed, that clears things up a lot!

Done.

**Comment by bgaydosrh** [comment]
Latest test build at http://file.bos.redhat.com/bgaydos/120219/cnv/cnv_users_guide/cnv-monitoring-vm-health.html

**Comment by bgaydosrh** [comment]
@alexxa - Are you blocked on QE'ing this one? Please let me know. Thanks, Bob

**Comment by bgaydosrh** [comment]
This PR still needs label for *enterprise-4.2*.

File: cnv/cnv_users_guide/cnv-monitoring-vm-health.adoc
Comment (ousleyp) [inline cnv/cnv_users_guide/cnv-monitoring-vm-health.adoc]: I would prefer if the intro for this assembly were its own concept module, with a brief intro sentence in the assembly instead.

File: cnv/cnv_users_guide/cnv-monitoring-vm-health.adoc
Comment (ousleyp) [inline cnv/cnv_users_guide/cnv-monitoring-vm-health.adoc]: I think this sentence is too long/wordy. Potential suggestion (I'm not sure if you need both examples or not):

"When a VirtualMachineInstance (VMI) fails, _liveness probes_ stop the VMI. Controllers such as `VirtualMachine` then spawn other VMIs, restoring virtual machine responsiveness."

(note that I think "liveness probes" should have emphasis since we're introducing them here -- correct me if I'm wrong about that)

File: cnv/cnv_users_guide/cnv-monitoring-vm-health.adoc
Comment (ousleyp) [inline cnv/cnv_users_guide/cnv-monitoring-vm-health.adoc]: - If this is the first place we're discussing readiness probes, I think that the term should be italicized (as with liveness probes)
- Why are Services and Endpoints capitalized?

File: modules/cnv-configure-liveness-readiness-probes.adoc
Comment (ousleyp) [inline modules/cnv-configure-liveness-readiness-probes.adoc]: This module needs to be split into three different procedure modules, because you can't have more than one procedure in a module. Each procedure needs an intro sentence above `.Procedure`. Each procedure should also include the `oc create -f` command at the end. (You'll need to change the bullets to numbers since there will be more than one step.)

**Review by ousleyp — CHANGES_REQUESTED** [review]
Some comments and structural changes. Please split up `cnv-configure-liveness-readiness-probes.adoc` into three different modules and then tag me for further review. Thanks!

File: modules/cnv-configure-liveness-readiness-probes.adoc
Comment (bgaydosrh) [reply modules/cnv-configure-liveness-readiness-probes.adoc]: Done. I have created three modules for this assembly:

include::modules/cnv-define-http-liveness-probe.adoc[leveloffset=+1]
include::modules/cnv-define-tcp-liveness-probe.adoc[leveloffset=+1]
include::modules/cnv-define-readiness-probe.adoc[leveloffset=+1]

I will issue a git rm to get rid of the old mega-module.

Added the intro sentences and the oc command step at the end of each.

**Comment by bgaydosrh** [comment]
> Some comments and structural changes. Please split up `cnv-configure-liveness-readiness-probes.adoc` into three different modules and then tag me for further review. Thanks!

See above for details.

File: cnv/cnv_users_guide/cnv-monitoring-vm-health.adoc
Comment (bgaydosrh) [reply cnv/cnv_users_guide/cnv-monitoring-vm-health.adoc]: Done.

File: cnv/cnv_users_guide/cnv-monitoring-vm-health.adoc
Comment (bgaydosrh) [reply cnv/cnv_users_guide/cnv-monitoring-vm-health.adoc]: Done.

File: cnv/cnv_users_guide/cnv-monitoring-vm-health.adoc
Comment (bgaydosrh) [reply cnv/cnv_users_guide/cnv-monitoring-vm-health.adoc]: Yes, it is first occurrence so I italicized.
Fixed capitals.

**Comment by bgaydosrh** [comment]
Hi @ousleyp - This is ready for your final review.

Here is the test build: http://file.bos.redhat.com/bgaydos/011020_2/cnv/cnv_users_guide/cnv-monitoring-vm-health.html

if this looks good I will merge and pick to 4.3. 

Thanks again for all the good feedback,

Bob

File: modules/cnv-define-readiness-probe.adoc
Comment (ousleyp) [inline modules/cnv-define-readiness-probe.adoc]: It might be better to include this info as callouts in the code block itself, but it might be okay as is.

File: modules/cnv-define-http-liveness-probe.adoc
Comment (ousleyp) [inline modules/cnv-define-http-liveness-probe.adoc]: Remove backticks

File: modules/cnv-define-readiness-probe.adoc
Comment (ousleyp) [inline modules/cnv-define-readiness-probe.adoc]: Remove backticks

File: modules/cnv-define-tcp-liveness-probe.adoc
Comment (ousleyp) [inline modules/cnv-define-tcp-liveness-probe.adoc]: As with the other module, it might be better to include this info in callouts in the code block, but I think it's also okay as is.

File: modules/cnv-define-tcp-liveness-probe.adoc
Comment (ousleyp) [inline modules/cnv-define-tcp-liveness-probe.adoc]: Remove backticks

File: modules/cnv-define-tcp-liveness-probe.adoc
Comment (ousleyp) [inline modules/cnv-define-tcp-liveness-probe.adoc]: This should be something like:
`$ oc create -f <file name>.yaml`

**Review by ousleyp — CHANGES_REQUESTED** [review]
Some minor issues need to be fixed, but otherwise it's looking much better. Thanks!

File: cnv/cnv_users_guide/cnv-monitoring-vm-health.adoc
Comment (bgaydosrh) [reply cnv/cnv_users_guide/cnv-monitoring-vm-health.adoc]: Changed to : "Use this procedure to create liveness and readiness probes
to monitor virtual machine health."

File: modules/cnv-about-liveness-readiness-probes.adoc
Comment (bgaydosrh) [reply modules/cnv-about-liveness-readiness-probes.adoc]: Gotcha. Done

File: modules/cnv-about-liveness-readiness-probes.adoc
Comment (bgaydosrh) [reply modules/cnv-about-liveness-readiness-probes.adoc]: Ugh! Done.

File: modules/cnv-define-http-liveness-probe.adoc
Comment (bgaydosrh) [reply modules/cnv-define-http-liveness-probe.adoc]: Done.

File: modules/cnv-define-http-liveness-probe.adoc
Comment (bgaydosrh) [reply modules/cnv-define-http-liveness-probe.adoc]: Done.

File: modules/cnv-define-readiness-probe.adoc
Comment (bgaydosrh) [reply modules/cnv-define-readiness-probe.adoc]: Done.

File: modules/cnv-define-readiness-probe.adoc
Comment (bgaydosrh) [reply modules/cnv-define-readiness-probe.adoc]: Done.

File: modules/cnv-define-readiness-probe.adoc
Comment (bgaydosrh) [reply modules/cnv-define-readiness-probe.adoc]: Good catch! Thanks. Done.

File: modules/cnv-define-http-liveness-probe.adoc
Comment (bgaydosrh) [reply modules/cnv-define-http-liveness-probe.adoc]: Done.

File: modules/cnv-define-readiness-probe.adoc
Comment (bgaydosrh) [reply modules/cnv-define-readiness-probe.adoc]: Done.

File: modules/cnv-define-tcp-liveness-probe.adoc
Comment (bgaydosrh) [reply modules/cnv-define-tcp-liveness-probe.adoc]: Done.

File: modules/cnv-define-http-liveness-probe.adoc
Comment (bgaydosrh) [reply modules/cnv-define-http-liveness-probe.adoc]: Done.

File: modules/cnv-define-readiness-probe.adoc
Comment (bgaydosrh) [reply modules/cnv-define-readiness-probe.adoc]: Done.

File: modules/cnv-define-tcp-liveness-probe.adoc
Comment (bgaydosrh) [reply modules/cnv-define-tcp-liveness-probe.adoc]: I'll keep that in mind for future... sounds like a good idea.

File: modules/cnv-define-tcp-liveness-probe.adoc
Comment (bgaydosrh) [reply modules/cnv-define-tcp-liveness-probe.adoc]: Done.

File: modules/cnv-define-tcp-liveness-probe.adoc
Comment (bgaydosrh) [reply modules/cnv-define-tcp-liveness-probe.adoc]: Done.

**Comment by bgaydosrh** [comment]
Test Build: http://file.bos.redhat.com/bgaydos/011520/cnv/cnv_users_guide/cnv-monitoring-vm-health.html

Ready for merge and pick to enterprise-4.3

**Pull request merged**
